### DISCUSSION TITLE
Configuring RawInput::predicted_dt to 0?

### DISCUSSION THREAD

**Discussion by barries** [emilk/egui#5353] [Q&A]
We're using `ctx.request_repaint_after()` to schedule an update, but have found a very surprising behavior where `egui` doesn't wait as long as requested, resulting in premature callbacks, causing many surprising unnecessary updates. For now, we've hacked a workaround (see below) by setting `RawInput::predicted_dt` to `0` in `raw_input_hook()`.

Is there a way to disable it, rather than work around it?

Separately, is the behavior documented anywhere?

Also, should this be disabled in reactive mode (and maybe in continuous mode--it's pretty likely to not guess right and cause extra repaints, IIUC)?

Thanks!

```rust
impl eframe::App for IotasoftUiApp {                                                    
    fn raw_input_hook(&mut self, _ctx: &egui::Context, raw_input: &mut egui::RawInput) {
        raw_input.predicted_dt = 0.0; // work around undocumented request_repaint_after() behavior that silently causes premature updates
    }                                                                                   
                                                                                        
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {             
        self.manage_polling();  // polls hardware, then sets self.next_poll_time
                                                                                        
        ... rendering code ...
                                                                                        
        let now = Instant::now();                                                       
        if now >= self.next_poll_time {                                                                                     
            self.ctx.request_repaint();                                                 
        } else {                                                                        
            let poll_remaining_time = self.next_poll_time - now;                                                  
            self.ctx.request_repaint_after(poll_remaining_time);                        
        }                                                                               
    }                                                                                   
}                                                                                       
```

**Comment by Platon7788** [comment]
Ran into exactly this and spent a while chasing it, so here is what is happening and the second half of the workaround, which is easy to miss.

**Why it fires early.** `ContextImpl::request_repaint_after` subtracts the predicted frame time before storing the delay:

```rust
if let Ok(predicted_frame_time) = Duration::try_from_secs_f32(viewport.input.predicted_dt) {
    // Make it less likely we over-shoot the target:
    delay = delay.saturating_sub(predicted_frame_time);
}
```

The reasoning is sound — you want the frame *on screen* at the deadline, so wake early enough to draw it. But `predicted_dt` is documented as "should be set to the expected time between frames when painting at vsync speeds", and **nothing in `egui-winit` ever assigns it**. It keeps its `1.0 / 60.0` default for the life of the process, whatever the application's real frame time is.

So on the winit backends the subtraction is a constant 16.67 ms:

- `request_repaint_after(1s)` sleeps 983 ms. Nobody notices.
- `request_repaint_after(33ms)` sleeps 16.6 ms — **the window runs at 60 fps while the code believes it asked for 30.**

That second line cost me a lot. Our frame-pacing module was inert for weeks, and two rounds of CPU conclusions drawn from it were nonsense, including a "cliff between 30 and 60 fps" that was just driver variance at a constant 60. Confirming it took one line: log `i.unstable_dt` next to the delay you asked for.

**The second half of the workaround.** `raw_input.predicted_dt = 0.0` fixes the scheduling, and then every animation stops. `InputState::begin_pass` does this:

```rust
let stable_dt = if requested_immediate_repaint_prev_frame {
    unstable_dt
} else {
    new.predicted_dt
};
```

If you schedule with `request_repaint_after` rather than `request_repaint`, `requested_immediate_repaint_prev_frame` is false every frame, so `stable_dt` **is** `predicted_dt` — which you just set to zero. Anything integrating `stable_dt` freezes.

The fix is to read `unstable_dt` instead, which is real wall time. Worth doing even without the hook: in our case `stable_dt` had been feeding a constant 16.67 ms into a download-rate meter, which was therefore wrong at every frame rate except 60, and into an animation clock, which ran ~60x too slow whenever the window dropped to a low background rate.

**Suggestions, in case any are useful:**

1. Document the subtraction on `request_repaint_after`. Right now `predicted_dt` says "can safely be left at its default value", which is true for its own purposes but not for anything that reads it back through `request_repaint_after` or `stable_dt`.
2. Have `egui-winit` set `predicted_dt` from the measured frame time. That makes the compensation do what it was designed to do instead of applying a fixed 16.67 ms to applications whose frames cost a millisecond.
3. Or skip the subtraction when the requested delay is much larger than a frame, so an explicit throttle means what it says.

Happy to open an issue or a PR for any of these if you have a preference for which.
